File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int freq[128] = {0};
        int left = 0, right = -1;  // to ensure that the silding window [left, right] has no character at first
        int len = s.size(), ans = 0;
        while (left < len) {
            if (right + 1 < len && freq[s[right + 1]] == 0) {
                freq[s[++right]]++;
            } else {
                freq[s[left++]]--;
            }
            ans = max(right - left + 1, ans);
        }
        return ans;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = 0
        d = defaultdict(int)
        l, r = 0, 0
        while l <= r < len(s):
            if d[s[r]] == 1:
                d[s[l]] -= 1
                l += 1
            else:
                d[s[r]] += 1
                r += 1
            ans = max(ans, r-l)
        return ans

        